infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/KubernetesRecipeParser.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.kubernetes.environment;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static java.lang.String.format;
+
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.client.KubernetesClientException;
+import java.io.ByteArrayInputStream;
+import java.util.List;
+import javax.inject.Inject;
+import org.eclipse.che.api.core.ValidationException;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.spi.environment.InternalRecipe;
+import org.eclipse.che.workspace.infrastructure.kubernetes.KubernetesClientFactory;
+
+/**
+ * Parses Kubernetes objects from recipe.
+ *
+ * <p>Note that this class can also parse OpenShift specific objects.
+ *
+ * @author Sergii Leshchenko
+ */
+public class KubernetesRecipeParser {
+  private final KubernetesClientFactory clientFactory;
+
+  @Inject
+  public KubernetesRecipeParser(KubernetesClientFactory clientFactory) {
+    this.clientFactory = clientFactory;
+  }
+
+  /**
+   * Parses Kubernetes objects from recipe.
+   *
+   * @param recipe that contains objects to parse
+   * @return parsed objects
+   * @throws IllegalArgumentException if recipe content is null
+   * @throws IllegalArgumentException if recipe content type is null
+   * @throws ValidationException if recipe content has broken format
+   * @throws ValidationException if recipe content non known objects
+   * @throws InfrastructureException when exception occurred during kubernetes client creation
+   */
+  public List<HasMetadata> parse(InternalRecipe recipe)
+      throws ValidationException, InfrastructureException {
+    String content = recipe.getContent();
```

Comment:
Note that it is `InternalRecipe` and it can not have a location.
`InternalRecipe` must already have content and it is resolved here https://github.com/eclipse/che/blob/master/wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/environment/InternalEnvironmentFactory.java#L93

Reply:
ok, tnx